### PR TITLE
Msgpack_read/write and some minor improvements

### PR DESCRIPTION
**Pull request by traversc**

### CODE REVIEW

File: DESCRIPTION [hunk 1]
```diff
@@ -1,8 +1,8 @@
 Package: RcppMsgPack
 Type: Package
 Title: 'MsgPack' C++ Header Files and Interface Functions for R
-Version: 0.2.2
-Date: 2018-05-06
+Version: 0.2.3
```
Comment (eddelbuettel) [inline DESCRIPTION:4]: I would call this 0.2.2.1 for the PR, and only go to 0.2.3 once we're ready for CRAN but no need to change it, I can do this here.

File: NAMESPACE [hunk 2]
```diff
@@ -3,4 +3,5 @@ importFrom("Rcpp", "evalCpp")
 export("msgpack_format", "msgpack_map", "msgpack_pack", "msgpack_simplify", "msgpack_unpack",
        "msgpackFormat", "msgpackMap", "msgpackPack", "msgpackSimplify", "msgpackUnpack",
        "msgpack_timestamp_encode", "msgpackTimestampEncode", "msgpack_timestamp_decode", "msgpackTimestampDecode",
+       "msgpack_write", "msgpackWrite", "msgpack_read", "msgpackRead",
```
Comment (eddelbuettel) [inline NAMESPACE:6]: I absolutely love our double convention :)  Others should follow our lead ;-)

File: src/RcppExports.cpp [hunk 3]
```diff
@@ -49,13 +49,13 @@ BEGIN_RCPP
 END_RCPP
 }
 // c_timestamp_encode
-RawVector c_timestamp_encode(double seconds, uint32_t nanoseconds);
+RawVector c_timestamp_encode(double seconds, u_int32_t nanoseconds);
```
Comment (eddelbuettel) [inline src/RcppExports.cpp:52]: I would have thought `uint32_t`  was more common / standard?  I guess you like how `u_` sticks out?

**Review by eddelbuettel — APPROVED** [review]
Looks solid. I did not dive into the cpp file as GH did not show it, and it is a bit weird that the one test file shows it as a complete change.  Oh well.  

Will merge once travis is through.

**Comment by traversc** [comment]
Pretty minor changes in C++.  I'll summarize:

* Use `msgpack::sbuffer` instead of `std::stringstream` as it allows direct access to the underlying data and seems to be marginally faster.  Also use a template for the buffer (although probably incorrectly labeled "stream"... )

* Use `std::vector<SEXP>` at first instead of List in the main `c_unpack` function, since you don't know ahead of time how many objects are in the message

* got rid of all the global variables used for temporary assignment ... seems like bad practice

**Comment by eddelbuettel** [comment]
Hm, trying to think if `std::vector<SEXP>` had bitten us before.  Seems like an obvious idea but I can't recall if we did hit potholes.  

I presume you had been running code like that your side for a while?   Rest sounds cool -- nice work, once again!

**Comment by traversc** [comment]
I haven't tested it extensively, but so far I haven't seen any problems.  I'm guessing there might be an issue with scope/deallocation of the underlying data?  Since I'm not returning or passing the vector around, perhaps we don't have to worry about that.

**Pull request merged**